### ISSUE TITLE
OpenJpegProcessor operating on low bit-depth images

### ISSUE THREAD

**Issue by adam-vessey** [medusa-project/cantaloupe#190]
There seems to be a [known issue with bit depths, due to the use of BMP intermediates](https://github.com/medusa-project/cantaloupe/blob/ffbfb5d41f80fd2b8beec155567938912048668d/src/main/java/edu/illinois/library/cantaloupe/processor/OpenJpegProcessor.java#L58-L60); however, it seems to describe the opposite situation. When testing in the wild, we have come across grayscale JPEG 2000 images with a bit-depth less than 8, which [OpenJPEG does not support with BMP](https://github.com/uclouvain/openjpeg/blob/9d0d1a0128c7a9324c9a935b2ded01deb4d660d4/src/bin/jp2/convertbmp.c#L866-L870)... this error makes its way into log files:

```
2018-01-03 16:14:10,613 ERROR [http-bio-8080-exec-13] o.r.C.Server [Slf4jLogger.java:230] An exception occurred writing the response entity
java.io.IOException: edu.illinois.library.cantaloupe.processor.ProcessorException: Unable to read the image header. (command output:
===========================================
The extension of this file is incorrect.
FOUND 5f1a. SHOULD BE .jp2
===========================================
imagetobmp: Unsupported precision: 2
[ERROR] Outfile /usr/local/fedora/tomcat/temp/cantaloupe-OpenJpegProcessor-2de6d695-0da1-42f8-86c7-1ad64d2acab1.bmp not generated
)
        at edu.illinois.library.cantaloupe.resource.ImageRepresentation.write(ImageRepresentation.java:125)
        at org.restlet.engine.adapter.ServerCall.writeResponseBody(ServerCall.java:513)
        at org.restlet.engine.adapter.ServerCall.sendResponse(ServerCall.java:457)
        at org.restlet.ext.servlet.internal.ServletCall.sendResponse(ServletCall.java:421)
        at org.restlet.engine.adapter.ServerAdapter.commit(ServerAdapter.java:191)
        at org.restlet.engine.adapter.HttpServerHelper.handle(HttpServerHelper.java:144)
        at org.restlet.ext.servlet.ServerServlet.service(ServerServlet.java:1117)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
        at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:950)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:408)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1070)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:611)
        at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: edu.illinois.library.cantaloupe.processor.ProcessorException: Unable to read the image header. (command output:
===========================================
The extension of this file is incorrect.
FOUND 5f1a. SHOULD BE .jp2
===========================================
imagetobmp: Unsupported precision: 2
[ERROR] Outfile /usr/local/fedora/tomcat/temp/cantaloupe-OpenJpegProcessor-2de6d695-0da1-42f8-86c7-1ad64d2acab1.bmp not generated
)
        at edu.illinois.library.cantaloupe.processor.OpenJpegProcessor.process(OpenJpegProcessor.java:419)
        at edu.illinois.library.cantaloupe.resource.ImageRepresentation.doWrite(ImageRepresentation.java:162)
        at edu.illinois.library.cantaloupe.resource.ImageRepresentation.write(ImageRepresentation.java:123)
        ... 27 common frames omitted
Caused by: javax.imageio.IIOException: Unable to read the image header.
        at com.sun.imageio.plugins.bmp.BMPImageReader.read(BMPImageReader.java:746)
        at javax.imageio.ImageReader.read(ImageReader.java:939)
        at edu.illinois.library.cantaloupe.processor.imageio.AbstractImageReader.read(AbstractImageReader.java:266)
        at edu.illinois.library.cantaloupe.processor.imageio.ImageReader.read(ImageReader.java:236)
        at edu.illinois.library.cantaloupe.processor.OpenJpegProcessor.process(OpenJpegProcessor.java:383)
        ... 29 common frames omitted
Caused by: java.lang.IllegalArgumentException: Invalid magic value for BMP file.
        at com.sun.imageio.plugins.bmp.BMPImageReader.readHeader(BMPImageReader.java:244)
        at com.sun.imageio.plugins.bmp.BMPImageReader.read(BMPImageReader.java:744)
        ... 33 common frames omitted
```

After this happens the servlet ends up in a non-functional state.

----

To work around this technical limitation of OpenJPEG, it appears that it may be possible to use PNG as an intermediate format. Ideally, we could only use PNG when necessary, so as to avoid the additional cycles (de)compressing PNGs. A [somewhat naive attempt has been made in this regard, with some success](https://github.com/medusa-project/cantaloupe/compare/dad9708796b1e27621c3000f1d96d34f8b2872cb...adam-vessey:snapshot-dynamic-openjpeg-intermediate).

Alternatively, it may be possible to put together an implementation at some level for ImageIO to use a "raw" dump of pixels from OpenJPEG, to similarly avoid the (de)compression hit.

**Comment by adolski** [comment]
Hi @adam-vessey,

Could one of the other intermediates substitute for the BMP+PNG duo? TIFF maybe? I don't remember why I chose BMP in the first place. But if not, then I think I would try to work around this similarly to how you are.

I would love to have an ImageIO plugin for OpenJPEG. This is only one of the many drawbacks to using `opj_decompress`.

**Comment by adam-vessey** [comment]
Something of conjecture, but I suspect BMP might have been selected as it is [one of the formats "guaranteed" to be supported by Image I/O](https://docs.oracle.com/javase/tutorial/2d/images/saveimage.html), and it's a fairly "raw" format (no additional (de)compression during de-/encoding directly tied to its use)... and doesn't have many other options tied to its use like TIFF can have.

... That said, I could see TIFF working well so long as OpenJPEG doesn't try to compress with LZW by default (quick experiment seems to indicate it does not). For what it's work, I looked at the other formats supported by OpenJPEG (`PBM|PGM|PPM|PNM|PAM|PGX|PNG|BMP|TIF|RAW|RAWL|TGA` in my version): the NetPBM formats (PBM/PGM/PPM/PNM/PAM) may also be suitable, but would need an Image I/O reader to fit into the present processing... a bit of searching around finds [the TwelveMonkeys library](https://github.com/haraldk/TwelveMonkeys/), which appears to have Image I/O plugins supporting NetPBM formats (and TIFF and more).

**Comment by adam-vessey** [comment]
It appears that [BMP _used_ to be the only output of OpenJPEG to which profiles were appropriately applied](https://github.com/uclouvain/openjpeg/issues/171); however, the issue is marked as resolved... a [bit of experimentation](https://gist.github.com/adam-vessey/ae4fa5baa02349384ce3382c7b4721a8) appears to support this: Where the "BMP truncation" does _not_ occur, the images appear to contain identical pixel content between the BMP and TIFF.

A [bit more experimentation indicates that OpenJPEG is slow generating the NetPBM](https://gist.github.com/adam-vessey/24f4f74c3546062c9926bc456243601c) formats... TIFF is slightly slower than BMP, while NetPBM was over double the time of BMP; though, this is just the time for getting output from OpenJPEG. I don't have a good harness at hand for timing the full response from Cantaloupe.

(an aside: `pnm` automatically selects the format from PBM, PGM and PPM, based on the colorspace of the image being transcoded)

Looks like TIFF might be the winner, appearing to hit all four criteria.

**Comment by adolski** [comment]
Thanks for looking into this, @adam-vessey. I agree that TIFF looks like the best candidate.

As far as scheduling a migration from BMP to TIFF, I think it's OK to do this in the `release/3.4` branch for inclusion in 3.4.1. The only issue is users who may have an older version of OpenJPEG installed, who upgrade Cantaloupe and then find that output doesn't look right anymore because their older opj_decompress doesn't respect ICC profiles in TIFF. But, it looks from the issue like this was fixed in 2.1.1, and OpenJPEG <2.2.0 was deprecated in 3.4, so hopefully they have upgraded by now, and if not, then they can just do so.

**Comment by adam-vessey** [comment]
I've taken [a shot at the implementation with TIFFs](https://github.com/medusa-project/cantaloupe/compare/develop...adam-vessey:openjpeg-tiff-intermediate)... in doing so, I found that it was necessary to move to using temporary files instead of the in-memory streaming of the intermediate files.

Some experimentation using stdout:
```
$ ls -l asdf.tif
lrwxrwxrwx 1 vagrant vagrant 11 Jan 31 18:26 asdf.tif -> /dev/stdout
$ opj_decompress -quiet -i 4-bit-grayscale-noise.jp2 -o asdf.tif > qwer.tif

$ echo $?
0
$ identify qwer.tif
qwer.tif TIFF 10000x10000 10000x10000+0+0 4-bit Grayscale DirectClass 50.08MB 0.000u 0:00.019
$ opj_decompress -quiet -i 4-bit-grayscale-noise.jp2 -o asdf.tif | cat > qwer.tif

TIFFAppendToStrip: Maximum TIFF file size exceeded.
[... previous line is repeated 10000 times ...]
$ echo ${PIPESTATUS[@]}
0 0
$ identify qwer.tif
[ ... no output at all ...]
$ echo $?
0
$ ls -l qwer.tif
-rw-rw-r-- 1 vagrant vagrant 8 Feb  1 16:30 qwer.tif
$ hexdump -C qwer.tif
00000000  4d 4d 00 2a 00 00 00 00                           |MM.*....|
00000008
```

I suspect the first case works since it can resolve the actual location of the file to obtain the random access libtiff requires, while the second fails to do so, despite the command returning success on the CLI. The contents of the file written _does_ appear to be the [first few bytes of a TIFF in big-endian order](https://web.archive.org/web/20150503034412/http://partners.adobe.com/public/developer/en/tiff/TIFF6.pdf).

Possibly went a little too deep while looking at this, but it appears that:
1. [OpenJPEG calls `TIFFOpen()`](https://github.com/uclouvain/openjpeg/blob/07d526e4cb93b6d2e1a52cf49e3d011d3d4caa11/src/bin/jp2/converttif.c#L635) (mode indicating  **w**rite and **b**ig-endian).
2. [`TIFFOpen()` calls `TIFFFdOpen()`](https://github.com/uclouvain/openjpeg/blob/3aaeea7ce8c2065d29483817b7fd07226601851c/thirdparty/libtiff/tif_unix.c#L250)
3. As per [libtiff's man pages](http://www.libtiff.org/man/TIFFOpen.3t.html):
    > The object associated with the file descriptor must support random access.

Thoughts?

**Comment by adolski** [comment]
That's a bummer. It's true that TIFF requires a random access writer. KakaduProcessor is using TIFF intermediates, but I guess kdu_expand gets around that by not using libtiff and writing into a memory buffer.

Have you observed the performance of this on-disk intermediate solution? I confess that I don't really like the idea of it and I assume that it's slow. But I wonder how slow.

The nice thing about BMP is that it streams efficiently from opj_decompress and is fast to read. So before going down that road, please tell me what you think about [this patch](https://github.com/adolski/openjpeg/commit/93dd6bd305074bf1dc1b0bc2f40f7b2e45b655a4) to the `imagetobmp()` function in OpenJPEG, which rescales all component sizes to 8 bits. A proper solution would be to write different types of BMPs depending on the component size, but each type will require different headers.

......

Now that I think about it, one of the benefits of writing the intermediates to disk is that it enables OpenJpegProcessor to work in Windows, which never occurred to me before. ...... I don't personally care about Windows, but maybe somebody out there does.

**Comment by adam-vessey** [comment]
Had some fun with JMH, and made it run over the `testProcessOnAllFixtures()` bit.

TL;DR: BMPs streaming vs temp files on a RAM disk seems to be comparable (a hit in the handling of files, but... is it appreciable?... falls within the error margins); also, TIFFs are comparatively _slow_ to process with OpenJPEG.

That said, I'm not sure about the OpenJPEG patch... if it was to be pushed upstream (behind a runtime CLI flag?), I could see making use of it; however, running a patched/forked version of OpenJPEG would be a show-stopper.

Getting somewhat back on target: Of primary concern is the the servlet being put into a non-functional state when these low bit-depth images are encountered (mentioned in the original report). We could live with it failing to serve for specific images, so long as it continued to serve later requests without requiring the servlet to be restarted... handling with files appears to fix this issue with the servlet failing to serve later requests. Actually serving images for low bit-depth images can be resolved later.

For actually serving them, we could approach a few different ways:
* get whatever options for BMP into OpenJPEG upstream
* determine the bit-depth of the source image in order to select which intermediate format to use
    * relatively constant hit on _every_ request, determining the bit-depth
        * could potentially be avoided by caching either the bit-depth or the intermediate format selected somewhere.
* fallback to using TIFF if the attempt to use BMP fails
    * should only impact those images which fail; however, this impact could be rather significant, effectively requiring OpenJPEG to decode these requests twice.
        * could potentially cache the used intermediate format to avoid the double-decoding behaviour?

----

"Benchmarking" (all just for comparison with our own results)


I've some made changes in the OpenJpegProcessorPerformance bit to run the processing method, and parameterized the intermediate format...

```
OpenJpeg as-is (testProcessOnAllFixtures does streaming BMP):
Benchmark                                                                           Mode  Cnt       Score        Error  Units
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  348743.704 ± 39208.655  us/op

OpenJpeg goes to files (testProcessOnAllFixtures is TIFF files):
Benchmark                                                                           Mode  Cnt       Score        Error  Units
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  909695.342 ± 103342.406  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesBMP                            avgt   10  485970.817 ± 134029.975  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesTIFF                           avgt   10  906578.403 ± 193400.689  us/op

OpenJpeg goes to files, on RAM disk:
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  869944.446 ± 35297.556  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesBMP                            avgt   10  368642.035 ± 56545.303  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesTIFF                           avgt   10  869244.253 ± 33824.379  us/op
```

Going to a _filesystem_ doesn't appear to be too much slower in and of itself; however, the caveat is that you may be limited based on where the filesystem lives and/or its caching settings... a RAM disk for the temp dir is essentially the best case? (when _not_ using a RAM disk, the temp directory was on a HDD in my benchmarking).

TIFFs _do_ seem remarkably slower here; however, wondering if having larger images on which to test processing change anything (the fixtures are all rather small): I generated a handful (30) of bigger JP2s with randomized dimensions centered around 8000, containing random noise, ~5 MiB files, each. To handle them, I had to add a cropping operation to the base testProcessOnAllFixtures; otherwise, was erroring due to heap space requirements.

```
Some big JP2s, to disk:
Benchmark                                                                           Mode  Cnt       Score        Error  Units
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  1363410.606 ± 231741.691  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesBMP                            avgt   10   768170.747 ± 283794.102  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesTIFF                           avgt   10  1140324.920 ± 436248.253  us/op

Some big JP2s, to RAM disk:
Benchmark                                                                           Mode  Cnt       Score        Error  Units
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  1045444.654 ± 193226.151  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesBMP                            avgt   10   619271.815 ±  61235.525  us/op
OpenJpegProcessorPerformance.testProcessOnAllFixturesTIFF                           avgt   10  1031965.433 ± 204010.862  us/o

Some big JP2s, streaming BMPs:
OpenJpegProcessorPerformance.testProcessOnAllFixtures                               avgt   10  602495.158 ± 49103.270  us/op
```

**Comment by adolski** [comment]
Whoa, I missed the part about OpenJpegProcessor ending up in a non-functional state. opj_decompress deletes the symlink when it errors out!

I've submitted a PR to change that: https://github.com/uclouvain/openjpeg/pull/1063

And my component-size-rescaling patch: https://github.com/uclouvain/openjpeg/pull/1064

The second one would effectively fix this issue (at least in the case of errors resulting from <8-bit images, not sure about others) and the first one would at least isolate the error to a single request.

As a quick fix, what if OpenJpegProcessor, in the event of this error, just recreated the symlink and tried again? When the symlink is missing I get:

```
java.lang.IllegalArgumentException: Invalid magic value for BMP file.
	at com.sun.media.imageioimpl.plugins.bmp.BMPImageReader.readHeader(BMPImageReader.java:248)
	at com.sun.media.imageioimpl.plugins.bmp.BMPImageReader.read(BMPImageReader.java:736)
	at java.desktop/javax.imageio.ImageReader.read(ImageReader.java:938)
	at edu.illinois.library.cantaloupe.processor.imageio.AbstractImageReader.read(AbstractImageReader.java:327)
	at edu.illinois.library.cantaloupe.processor.imageio.ImageReader.read(ImageReader.java:320)
	at edu.illinois.library.cantaloupe.processor.OpenJpegProcessor.process(OpenJpegProcessor.java:340)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.doWrite(ImageRepresentation.java:164)
	at edu.illinois.library.cantaloupe.resource.ImageRepresentation.write(ImageRepresentation.java:130)
```

So `OpenJpegProcessor.process()` would have to catch that exception, try to recreate the symlink, and call itself again (up to n times to prevent an infinite loop). That's pretty ugly, but it looks like the alternatives are just as ugly in some way or other?

Thanks for doing all of that testing. Good stuff to know.

**Comment by adam-vessey** [comment]
Having the symlink go away and be recreated between separate threads sounds like a race-condition in the making: Even if the creation is synchronized, if things can be pulled out from under you at any time by another thread erroring, things could get messy... to loop would occasionally account for it, but... could possibly behave better if each processor instance had its own symlink, either allocated per instance, or drawn from a pool of symlinks (just need to make sure that it exists when returning to the pool; is a pool worth it, given added complexity?)?

**Comment by RussellMcOrmond** [comment]
I believe the cleanest thing to do would be to add a patch to openjpeg to explicitly handle using openjpeg via pipes.  Many tools allow a single dash to be used to indicate to use stdin when indicated for input, and stdout if it is indicated for output.    This makes what is being requested be more explicit, and makes the logic within openjpeg clear (these are not files on disk, but filehandles passed to the process).

**Comment by adolski** [comment]
Symlink-per-instance ought to work fine. `process()` will need to catch and rethrow Throwable to make sure they get cleaned up.
